include/pybind11/cast.h

```diff
@@ -2198,6 +2198,17 @@ object object_api<Derived>::call(Args &&...args) const {
 
 PYBIND11_NAMESPACE_END(detail)
 
+
+/** \ingroup python_builtins
+    \rst
+    Return the registered type object for a C++ class, given as a template parameter.
+    py::type<T>() returns the Python type object previously registered for T.
+\endrst */
+template<typename T>
+handle type() {
+    return detail::get_type_handle(typeid(T), true);
```

Comment:
Is there a way to have this return something like `py::none()` in C++ when the user doesn't want a fail-fast?

Reply:
Not being able to convert a type is something that doesn't really exist in Python, so I selected the error throwing version of `get_type_handle`, but I think simply changing this to false would return none. Is that better than error checking? I don't think pybind11 is close to working / plans to work with error catching off?

Reply:
A C++ type could not (yet) be exposed, right?
But I think throwing an error should definitely be the default. Maybe the other option could/should stay in `detail`, even, and a try-catch could be the proposed public solution?

Reply:
But then yes, it's more efficient, so it could be possible. We mainly need to see that the API makes sense, indeed.

Another possible overload, btw: should we also have a `py::type(py::handle)` or so? Similar to all the `py::cast` variants?

Reply:
Going the other way is not possible, you can't return a C++ type, but having `py::handle py::type(py::handle a)` might make sense, as a way to do python's `type(a)` in C++.

Reply:
```cpp
inline handle type(handle h) {
    PyObject* obj = (PyObject *) Py_TYPE(h.ptr());
    return handle(obj);
}
```

works, would that make sense to add?

Reply:
(Based on https://github.com/python/cpython/blob/eaa551702d80fd67219c48ee6a13ffb571ca360b/Objects/typeobject.c#L990-L1001 )

Reply:
Yes, that's what I meant! Not to C++ ofc; sorry to be unclear!
Even better, maybe: `return obj.get_type()`?

Reply:
Free functions are generally better than member functions in C++ - simpler/more composable, etc. Maybe adding this then eventually deprecating / removing `h.get_type()`?

Reply:
It would simplify the mixin, and keep the API closer to Python. `h.str()` is already deprecated in favor of `py:str()`.

Reply:
`handle::get_type` does make sense to me, so I'm not sure if it's worth deprecating. If anything, I'd say to maybe implement `py::type(handle)` through `handle::get_type()`, just as a convenience, Python-like function.

Reply:
It is only used internally, not publicly in any docs or tests, so to me it seems it would read better (more python like) and simplify the mixin.

For example, randomly picking one of the 10 places it's used:

```cpp
if (!obj.get_type().is(matrix_type))
// Becomes
if (!py::type(obj).is(matrix_type))
```
For comparison, the Python would be:

```python
if not type(obj) is matrix_type
```

Reply:
OK, that's a good argument! I think I am/can be convinced, then. I just don't want to be the only one to mandate an API deprecation.

@rwgk, @bstaletic, @EricCousineau-TRI?

Reply:
We can also move the discussion into an issue, though I do need to come up with an implementation for `py::type(h)` - either the one above, or relying on `h.get_type()` (unless we also wait for that).

Reply:
Yeah, sorry, with this amount of comments from me, nothing ever gets merged. What about merging this PR (I believe everyone was in favor, anyway), and moving the other proposal to a next PR?